wasm/keystore.ts

```diff
@@ -0,0 +1,282 @@
+    name: string;
+    version: number;
+    activeAccounts: ActiveAccount[];
+  }
+
+  export interface IKeyStore {
+    // Check if wallet id exists
+    hasWallet(id: string): Promise<boolean>;
+
+    // Load a wallet by wallet id
+    load(id: string): Promise<Wallet>;
+
+    // Load all wallets
+    loadAll(): Promise<Wallet[]>;
+
+    // Import a wallet by mnemonic, name, password and initial active accounts (from coinTypes)
+    import(
+      mnemonic: string,
+      name: string,
+      password: string,
+      coins: CoinType[]
+    ): Promise<Wallet>;
+
+    // Import a wallet by private key, name and password
+    importKey(
+      key: Uint8Array,
+      name: string,
+      password: string,
+      coin: CoinType
+    ): Promise<Wallet>;
+
+    // Import a Wallet object directly
+    importWallet(wallet: Wallet): Promise<void>;
+
+    // Add active accounts to a wallet by wallet id, password, coin
+    addAccounts(
+      id: string,
+      password: string,
+      coins: CoinType[]
+    ): Promise<Wallet>;
+
+    // Get private key of an account by wallet id, password, coin and derivation path
+    getKey(
+      id: string,
+      password: string,
+      account: ActiveAccount
+    ): Promise<PrivateKey>;
+
+    // Delete a wallet by wallet id and password.
+    delete(id: string, password: string): Promise<void>;
+
+    // Export a wallet by wallet id and password, returns mnemonic or private key
+    export(id: string, password: string): Promise<string | Uint8Array>;
+  }
+
+  export interface IKeyStoreStorage {
+    get(id: string): Promise<Wallet | null>;
+    set(id: string, wallet: Wallet): Promise<void>;
+    loadAll(walletIdsKey: string): Promise<Wallet[]>;
+    delete(id: string, password: string): Promise<void>;
+  }
+
+  export class DefaultImpl implements IKeyStore {
+    private readonly core: WalletCore;
+    private readonly storage: IKeyStoreStorage;
+    private readonly walletIdsKey: string;
+
+    constructor(
+      core: WalletCore,
+      walletIdsKey: string,
+      storage: IKeyStoreStorage
+    ) {
+      this.core = core;
+      this.storage = storage;
+      this.walletIdsKey = walletIdsKey;
+    }
+
+    hasWallet(id: string): Promise<boolean> {
+      return this.storage
+        .get(id)
+        .then((wallet) => {
+          return wallet !== undefined;
+        })
+        .catch((error) => {
+          return false;
+        });
+    }
+
+    load(id: string): Promise<KeyStore.Wallet> {
+      return this.storage.get(id).then((wallet) => {
+        if (wallet === null) {
+          throw KeyStore.Error.WalletNotFound;
+        }
+        return wallet;
+      });
+    }
+
+    loadAll(): Promise<Wallet[]> {
+      return this.storage.loadAll(this.walletIdsKey);
+    }
+
+    delete(id: string, password: string): Promise<void> {
+      return this.storage.delete(id, password);
+    }
+
+    mapWallet(storedKey: StoredKey): KeyStore.Wallet {
+      let json = storedKey.exportJSON();
```

Comment:
Better to use const here since _json_, and _wallet_ are not updated, same on mapStoredKey